src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java

```diff
@@ -0,0 +1,277 @@
+import static org.folio.inventory.consortium.model.ConsortiumEnumStatus.ERROR;
+import static org.folio.inventory.dataimport.util.DataImportConstants.UNIQUE_ID_ERROR_MESSAGE;
+
+public class ConsortiumInstanceSharingHandler implements AsyncRecordHandler<String, String> {
+
+  private static final Logger LOGGER = LogManager.getLogger(DataImportKafkaHandler.class);
+
+  private static final String OKAPI_TOKEN_HEADER = "x-okapi-token";
+  private static final String OKAPI_URL_HEADER = "x-okapi-url";
+  private static final String OKAPI_TENANT_HEADER = "x-okapi-tenant";
+
+  private final Vertx vertx;
+  private final Storage storage;
+  private final KafkaConfig kafkaConfig;
+  private final Map<String, KafkaProducer<String, String>> producerList = new HashMap<>();
+  protected static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().setSerializationInclusion(JsonInclude.Include.NON_NULL)
+    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    .configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
+
+  public ConsortiumInstanceSharingHandler(Vertx vertx, Storage storage, KafkaConfig kafkaConfig) {
+    this.vertx = vertx;
+    this.storage = storage;
+    this.kafkaConfig = kafkaConfig;
+  }
+
+  @Override
+  public Future<String> handle(KafkaConsumerRecord<String, String> record) {
+    try {
+      Promise<String> promise = Promise.promise();
+      String consortiumId = record.key();
+      LOGGER.info("handle :: CONSORTIUM_INSTANCE_SHARING_INIT from consortiumId {}", consortiumId);
+      SharingInstance sharingInstance = Json.decodeValue(record.value(), SharingInstance.class);
+
+      String instanceId = sharingInstance.getInstanceIdentifier().toString();
+      Map<String, String> kafkaHeaders = KafkaHeaderUtils.kafkaHeadersToMap(record.headers());
+
+      LOGGER.info("Event CONSORTIUM_INSTANCE_SHARING_INIT has been received for instanceId: {}, sourceTenant: {}, targetTenant: {}",
+        instanceId, sharingInstance.getSourceTenantId(), sharingInstance.getTargetTenantId());
+      String tenantId = kafkaHeaders.get(OKAPI_TENANT_HEADER);
+
+      Context targetTenantContext = EventHandlingUtil.constructContext(sharingInstance.getTargetTenantId(),
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection targetInstanceCollection = storage.getInstanceCollection(targetTenantContext);
+
+      Context sourceTenantContext = EventHandlingUtil.constructContext(sharingInstance.getSourceTenantId(),
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection sourceInstanceCollection = storage.getInstanceCollection(sourceTenantContext);
+
+      LOGGER.info("handle :: checking is instance {} exists on tenant {}", instanceId, sharingInstance.getTargetTenantId());
+      getInstanceById(instanceId, sharingInstance.getTargetTenantId(), targetInstanceCollection)
+        .onFailure(failure -> {
```

Comment:
The chain of handlers here is too long(all the way to the end of this method). Please break it out into private methods to improve readability.

Reply:
Thanks, It will be fixed in scope of [MODINV-870](https://issues.folio.org/browse/MODINV-870 )